### ISSUE TITLE
change(ios): better handling for keyboard stability after background restoration

### ISSUE THREAD

**Issue by jahorton** [keymanapp/keyman#11259]
After a bit more searching... maybe I've found something this time?

- https://developer.apple.com/documentation/webkit/wknavigationdelegate/1455639-webviewwebcontentprocessdidtermi

We'd need to test and see if this is being called when our app is restored in this scenarios to be sure it's tied to the behavior that's been seen, of course.

Ref: https://github.com/ionic-team/capacitor/discussions/5488#Potential_solution

> The thing we need to be sensitive to is that when a WebView background termination occurs, webViewWebContentProcessDidTerminate() is invoked when the app is coming back into the foreground  [...]

Noting other issues from ionic and from react, they've seen cases where the WebView is 100% crashed, showing a blank page.  We haven't had _that_ issue, though we also aren't navigating across pages or doing other fancy web-server stuff on the back-end - it's just one page for us... albeit with a WebWorker backend and linked-in JS scripts.  There's a fair shot that iOS's WebView is doing its best to restore our page after such a termination but is coming up short.

- Example `react` issue: https://github.com/react-native-webview/react-native-webview/issues/3062

One concerning thing here, though - these are big teams that are struggling to get this scenario right if this is indeed a match for our circumstances.

_Originally posted by @jahorton in https://github.com/keymanapp/keyman/issues/10952#issuecomment-2065585146_

**Comment by jahorton** [comment]
> Noting other issues from ionic and from react, they've seen cases where the WebView is 100% crashed, showing a blank page. We haven't had _that_ issue[...]

Or have we?
- #11202

**Comment by mcdurdin** [comment]
I've also observed blank webview sporadically on ios.

**Comment by jahorton** [comment]
I'm beginning to suspect that this is actually quite related to the following:
- #12213 
- #13102.   

We've seen before from other issues that `WKWebView` can have odd issues when being restored after being backgrounded in app:
- #6542
- #2184
- #2292

And our fix - #10952 - seemed to knock those in-app issues out.  Of course, said fix was only for in-app... meaning we never made a similar fix for the system keyboard.  What if the first two issues, plus the reported occasional blank keyboard, are how similar problems surface for the system keyboard?

**Comment by jahorton** [comment]
From [this link](https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#oncontentprocessdidterminate) found via the React issue linked above:

> For instance, since iOS WebViews are not included in the total RAM of the app, they can be terminated independently of the app to liberate memory for new apps the user is opening. It's not unexpected to have WebViews get terminated after a while in the background.

One of the reasons I'm drawing the inferences that I am:  when I search for reasons that a WebView would fail to load a file or a script, almost without exception the responses will point to [`WKWebView.loadFileUrl(_, allowingReadAccessTo:)`](https://developer.apple.com/documentation/webkit/wkwebview/loadfileurl(_:allowingreadaccessto:)).  This search-issue is maintained even when I'm trying every keyword I can think of to indicate "even when the file's within the 'read access' zone."

So, the thought occurred to me - if the WebView process gets terminated... what happens to any special permissions settings it was given?  Are those reliably maintained by the replacement process?

**Comment by jahorton** [comment]
Regarding #12213:  

- in the original post:
    > [...] the bar disappears after a while and is replaced with the Keyman logo.

    Note:  **_after a while_**.
- in the final comment from the user on their Community site posting:
    >  Interestingly, the suggestion bar came back after a few days of use, all by itself.

Regarding #11202:

>  [...] 4. Typed a text on the Search bar. 5. Locked the mobile device by pressing Lock key from the mobile. 6. Unlock the mobile. Here, I noticed that the Keyman Keyboard was no longer visible in the Keyboard panel.

Note that locking a mobile device has been noted to contribute to other known cases of WebView process termination.

**Comment by jahorton** [comment]
I'm having difficulty attempting to perform the setup needed for a good repro on my personal device:  iOS is very often replacing the old instance of the extension if I hop out of an app and back in after even just a few seconds.  Interestingly, debugging via desktop Safari will keep the inspector view active _until_ I re-enter the app - it's only discarding it upon re-entry.

**Comment by jahorton** [comment]
Hmm.  I wonder if there's a slim chance that #12216 might have contributed to this in a manner that's now been fixed.  

Before #13169 landed, the memory leak would have kept old `WebView` instances alive, though disconnected, until the app extension itself was forcibly shut down.  But... would it still act the same under low memory conditions?  If the OS started trying to conserve and reuse resources before abandoning hope and terminating the extension, perhaps that could be the re-use case that previously exhibited the symptoms?

Keep in mind, WebKit content uses its own memory space, so by the time that _our_ extension is hitting its memory constraints from simply keeping the _connections_ to Safari / WebKit content alive... just how much memory was being tied up in the preserved WKWebView instances?  An issue which is now _fixed_ due to #13169.

> iOS is very often replacing the old instance of the extension if I hop out of an app and back in after even just a few seconds. [...]

At this point, I'm feeling like that "very often" is now actually an "always".  It doesn't even need to be "after even just a few seconds" - I can stop editing in Notes, then immediately re-tap the text area to resummon the keyboard - it'll still produce a new instance of our system keyboard.   _Maaaaybe_ iOS is being extra-cautious about something installed via TestFlight, but I'd figure **mcdurdin** above would have also been using TestFlight versions.  Or perhaps iOS is extra-cautious when there's an active debugger session attached?

I'm wondering if I'm just missing something regarding how to reproduce the scenarios that led to all of these reports... or if there's actually a slim chance that these were all side effects of something we just recently fixed.

**Comment by jahorton** [comment]
#13312 may also be associated with the analyses above.  Granted, it would be a bit odd for the main Web-engine script to fail to load but not ios-host.js to fail... though the Web-engine script is certainly much larger, thus more prone to being affected by memory-constraint issues.

**Comment by jahorton** [comment]
Reviewing the open issues linked above:
- #12213 
- #13102 
- #11202 
- #13312

At least a week later, it is still the case that none of these have any recorded Sentry events for releases since #13169 merged.  Outside of local dev builds (in particular, a case _intended_ to force-repro an error with similar format), the latest 18.0 build seen for any is for #12213:  `release@18.0.181-alpha`.

**Comment by mcdurdin** [comment]
> At least a week later, it is still the case that none of these have any recorded Sentry events for releases since [#13169](https://github.com/keymanapp/keyman/pull/13169) merged.

How many users do we have on 18.0 beta on iOS? I am guessing not many. We need to wait until the fixes hit stable to have any confidence on those measurements.

**Comment by jahorton** [comment]
> > At least a week later, it is still the case that none of these have any recorded Sentry events for releases since [#13169](https://github.com/keymanapp/keyman/pull/13169) merged.
> 
> How many users do we have on 18.0 beta on iOS? I am guessing not many. We need to wait until the fixes hit stable to have any confidence on those measurements.

Fair enough.  There's a reason I haven't closed this issue yet, after all - we do want more confidence here.

At this moment, our status.keyman.com dashboard shows literally 0 Sentry events for any 18.0 beta version.  Again, not enough to claim sufficient confidence, but still possibly noteworthy down the road.

**Comment by jahorton** [comment]
Another useful related resource: https://nevermeant.dev/handling-blank-wkwebviews/#killed-after-being-rendered
